src/test/rpc/ShardArchiveHandler_test.cpp

```diff
@@ -49,196 +49,207 @@ class ShardArchiveHandler_test : public beast::unit_test::suite
+        }
 
-                auto& session{handler->sqliteDB_->getSession()};
-                soci::rowset<soci::row> rs =
-                    (session.prepare << "SELECT * FROM State;");
+        {
+            std::lock_guard<std::mutex> lock(handler->m_);
 
-                uint64_t pos = 0;
-                for (auto it = rs.begin(); it != rs.end(); ++it, ++pos)
-                {
-                    BEAST_EXPECT(it->get<int>(0) == dl[pos].first);
-                    BEAST_EXPECT(it->get<std::string>(1) == dl[pos].second);
-                }
+            auto& session{handler->sqliteDB_->getSession()};
+            soci::rowset<soci::row> rs =
+                (session.prepare << "SELECT * FROM State;");
 
-                BEAST_EXPECT(pos == dl.size());
+            std::uint64_t pos = 0;
+            for (auto it = rs.begin(); it != rs.end(); ++it, ++pos)
+            {
+                BEAST_EXPECT(it->get<int>(0) == dl[pos].first);
+                BEAST_EXPECT(it->get<std::string>(1) == dl[pos].second);
             }
 
-            handler->release();
+            BEAST_EXPECT(pos == dl.size());
         }
 
-        // Destroy the singleton so we start fresh in
-        // the next testcase.
-        RPC::ShardArchiveHandler::instance_.reset();
+        handler->release();
     }
 
+    // Test the shard downloading module by initiating
+    // and completing ten downloads and verifying and
+    // the contents of the filesystem and the handler's
+    // archives.
     void
-    testStateDatabase3()
+    testDownloadsAndFileSystem()
     {
-        testcase("testStateDatabase3");
+        testcase("testDownloadsAndFileSystem");
 
-        {
-            beast::temp_dir tempDir;
+        beast::temp_dir tempDir;
 
-            auto c = jtx::envconfig();
-            auto& section = c->section(ConfigSection::shardDatabase());
-            section.set("path", tempDir.path());
-            section.set("max_size_gb", "100");
-            c->setupControl(true, true, true);
+        auto c = jtx::envconfig();
+        auto& section = c->section(ConfigSection::shardDatabase());
+        section.set("path", tempDir.path());
+        section.set("max_size_gb", "100");
+        section.set("ledgers_per_shard", "256");
+        c->setupControl(true, true, true);
 
-            jtx::Env env(*this, std::move(c));
-            auto handler = RPC::ShardArchiveHandler::getInstance(
-                env.app(), env.app().getJobQueue());
-            BEAST_EXPECT(handler);
+        jtx::Env env(*this, std::move(c));
 
-            BEAST_EXPECT(handler->init());
+        std::uint8_t const numberOfDownloads = 10;
 
-            auto server = createServer(env);
-            auto host = server.local_endpoint().address().to_string();
-            auto port = std::to_string(server.local_endpoint().port());
-            server.stop();
+        for (int i = 0; i < env.app().getShardStore()->ledgersPerShard() *
```

Comment:
It would be great if there were a comment explaining why you need to do so many `env.close()` calls before proceeding.  Just because we can see that it's happening doesn't mean your readers understand the motivation.

Reply:
Done